src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -613,12 +573,10 @@ DatabaseShardImp::onStop()
     if (shards_.empty())
         return;
 
-    // Notify and destroy shards
-    for (auto& e : shards_)
-    {
-        if (e.second.shard)
-            e.second.shard->stop();
-    }
+    // Notify shards to stop
+    for (auto const& e : shards_)
+        e.second->stop();
```

Comment:
Thank you for the deep review of the shutdown process! Yes, `DatabaseShard` owns `Shard` S. The `Shard` class is a private implementation that only DatabaseShard and its child, TaskQueue, touch. `DatabaseShard::OnStop` releases its ptrs and following the return, its child, `TaskQueue` releases any ptrs it might have. This works, but it can be improved. I'll implement `Stoppable::onChildrenStopped`, no need to poll. Upon being called, all `DatabaseShard` children should have released ptrs and stopped, otherwise something went wrong and it will be logged.

It would be nice to have Shard derive from Stoppable, but unfortunately the Stoppable class doesn't support removal of children.